indico/modules/users/client/js/react/components/ICSCalendarLink.jsx

```diff
@@ -0,0 +1,135 @@
+  const [option, setOption] = useState(null);
+  const [open, setOpen] = useState(false);
+
+  const copyButton = (
+    <Button
+      icon="copy"
+      onClick={async () => {
+        await navigator.clipboard.writeText(option.url);
+        setCopied(true);
+      }}
+    />
+  );
+
+  const fetchURL = async queryParams => {
+    try {
+      const {
+        data: {url: signedURL},
+      } = await indicoAxios.post(signURL(), snakifyKeys({endpoint, urlParams, queryParams}));
+      return signedURL;
+    } catch (error) {
+      handleAxiosError(error);
+    }
+  };
+
+  const trigger = (
+    <Dropdown
+      {...restProps}
+      icon={null}
+      trigger={
+        <Button icon size="small">
+          <Icon name="calendar alternate outline" />
+          <Icon name="caret down" />
+        </Button>
+      }
+      pointing="top right"
+    >
+      <Dropdown.Menu>
+        <Dropdown.Header>
+          <Translate>Synchronize with your calendar</Translate>
+        </Dropdown.Header>
+        <Dropdown.Divider />
+        {options.map(({key, text, queryParams}) => (
+          <Dropdown.Item
+            key={key}
+            text={text}
+            onClick={async () => {
+              setOption({
+                text,
+                url: await fetchURL(queryParams),
+              });
+              setOpen(true);
+            }}
+          />
+        ))}
+      </Dropdown.Menu>
+    </Dropdown>
+  );
+
+  return (
+    <Popup
+      trigger={trigger}
+      position="left center"
+      open={open}
+      popperDependencies={[copied]}
+      on="click"
+      onOpen={async () => {
+        setCopied(false);
+      }}
+      onClose={() => {
+        setOpen(false);
+      }}
+      wide
+    >
+      <Header
+        content={Translate.string('Sync with your calendar')}
+        subheader={option && <Label color="blue">{option.text}</Label>}
+      />
+      <Popup.Content>
+        <p>
+          <Translate>
+            You may copy-paste the following URL into your scheduling application. Contents will be
+            automatically synchronized.
+          </Translate>
+        </p>
+        <Input
+          placeholder={Translate.string('Loading…')}
+          readOnly
+          loading={!option}
+          size="mini"
+          fluid
+          value={option ? option.url : ''}
+          action={option && navigator.clipboard && copyButton}
+        />
+        {copied && (
+          <Grid centered>
+            <Grid.Row>
+              <Label pointing="above" color="teal">
+                <Translate>Copied!</Translate>
+              </Label>
+            </Grid.Row>
+          </Grid>
+        )}
+      </Popup.Content>
+    </Popup>
+  );
+}
+
+ICSCalendarLink.propTypes = {
+  endpoint: PropTypes.string.isRequired,
+  urlParams: PropTypes.objectOf(PropTypes.string),
+  options: PropTypes.arrayOf(
+    PropTypes.shape({
+      key: PropTypes.string,
+      text: PropTypes.string,
```

Comment:
`isRequired`? (same for key)

Reply:
ESLint doesn't complain about missing `isRequired` within `.shape({})`. I never add those because of that...

Reply:
I guess it only affects the actual proptypes checking at runtime, whether it'll complain about it being undefined or not.